File: C++/pico-boilerplate/pages/format_sd.cpp
//
// Created by Petr Jurásek on 26.06.2024.
//

#include "format_sd.h"

#include "../config/config.h"
#include "../config/Display.h"
#include "../config/sd_card_manager.h"
#include "../graphics/page_elements.h"
#include "../config/setup.h"

bool render_format_sd(){
    Display& display = Display::getInstance();
    pimoroni::ST7789& driver = display.getDriver();
    pimoroni::PicoGraphics_PenRGB332& graphics = display.getGraphics();

    sd_card_manager* sd_card_manager = sd_card_manager::get_instance();
    Buttons& buttons = Buttons::getInstance();

    graphics.set_pen(Colors::BLACK);
    graphics.clear();
    graphics.set_pen(Colors::WHITE);

    graphics.text("Format SD card", pimoroni::Point{2, 0}, 236, 2);
    graphics.set_pen(Colors::RED);
    graphics.text("WARNING: This will delete all data on the SD card!", pimoroni::Point{2, 20}, 236, 2);
    graphics.text("Press Y to confirm", Point{2, DISPLAY_HEIGHT - 40}, 236, 2);
    graphics.set_pen(Colors::GREEN);
    graphics.text("Press X to cancel", Point{2, DISPLAY_HEIGHT - 20}, 236, 2);
    driver.update(&graphics);

    while(!buttons.is_button_y_pressed()){
        if (buttons.is_button_x_pressed()){
            return false;
        }
    }

    sd_card_manager->format();
    initial();
    return true;
}

File: MicroPython/pages/format_sd.py
from utime import sleep
from config.config import Display, Colors
from config.buttons import button_x, button_y
from graphics import page_elements
from config.sdcard_manager import SDCardManager


def render_format_sd():
    """Renders the page for formatting the SD card."""

    display = Display()

    page_elements.clear_fast()
    display().set_pen(Colors.WHITE)
    display().text("Format SD card", 2, 0, 236, 2)
    display().set_pen(Colors.RED)
    display().text("Are you sure you want to format the SD card?", 2, 20, 236, 2)
    display().text("Press Y to confirm", 2, display.y_max - 20, 236, 2)
    display().update()

    while not button_y.read():
        if button_x.read():
            return False
        sleep(0.1)

    return format_sd()


def format_sd():
    """Formats the SD card."""

    sd = SDCardManager()

    if sd.is_mounted():
        sd.format_sd()

        return True

    return False
